File: common/src/test/java/DahRealPanda/plugins/colonyweb/util/TextTest.java
package DahRealPanda.plugins.colonyweb.util;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * {@link Text}'s string helpers, which every scanner leans on to turn raw MineColonies ids and
 * translation keys into something a dashboard can show.
 */
class TextTest {

    @Nested
    @DisplayName("pathOf")
    class PathOf {

        @Test
        @DisplayName("returns the part after the namespace")
        void stripsNamespace() {
            assertEquals("blockhutbuilder", Text.pathOf("minecolonies:blockhutbuilder"));
        }

        @Test
        @DisplayName("returns an unqualified id unchanged")
        void keepsUnqualified() {
            assertEquals("oak_planks", Text.pathOf("oak_planks"));
        }

        @Test
        @DisplayName("null becomes empty rather than throwing")
        void nullIsEmpty() {
            assertEquals("", Text.pathOf(null));
        }

        @ParameterizedTest(name = "\"{0}\" -> \"{1}\"")
        @DisplayName("handles ids with no path, no namespace, or extra colons")
        @CsvSource(delimiterString = "|", value = {
                "minecolonies:|",
                ":shingle|shingle",
                "domum_ornamentum:shingle#a1b2|shingle#a1b2",
                // Only the first colon separates; the rest belong to the path.
                "a:b:c|b:c",
        })
        void edgeCases(String input, String expected) {
            assertEquals(expected == null ? "" : expected, Text.pathOf(input));
        }
    }

    @Nested
    @DisplayName("humanize")
    class Humanize {

        @ParameterizedTest(name = "\"{0}\" -> \"{1}\"")
        @CsvSource(delimiterString = "|", value = {
                "oak_planks|Oak Planks",
                "blockHutBuilder|Block Hut Builder",
                "stone-age|Stone Age",
                "supplyship.blueprint|Supplyship",
                "citizen/builder|Builder",
                "level4|Level 4",
                "4level|4 Level",
                "already Spaced|Already Spaced",
        })
        void readableNames(String input, String expected) {
            assertEquals(expected, Text.humanize(input));
        }

        @ParameterizedTest
        @DisplayName("blank input yields blank output")
        @NullAndEmptySource
        @ValueSource(strings = {"   ", "\t"})
        void blankStaysBlank(String input) {
            assertEquals("", Text.humanize(input));
        }

        @Test
        @DisplayName("collapses runs of separators instead of emitting empty words")
        void collapsesSeparators() {
            assertEquals("Oak Planks", Text.humanize("oak___planks"));
        }

        @Test
        @DisplayName("a trailing separator does not leave trailing whitespace")
        void noTrailingSpace() {
            assertEquals("Oak", Text.humanize("oak_"));
        }
    }

    @Nested
    @DisplayName("looksLikeKey")
    class LooksLikeKey {

        @ParameterizedTest
        @DisplayName("dotted and namespaced ids look like keys")
        @ValueSource(strings = {
                "com.minecolonies.research.technology.stone.name",
                "minecolonies:blockhutbuilder",
                "a.b",
        })
        void keys(String input) {
            assertTrue(Text.looksLikeKey(input));
        }

        @ParameterizedTest
        @DisplayName("prose and bare words do not")
        @ValueSource(strings = {
                "Stone Age",
                "Builder",
                // A leading dot or colon is not a separator between segments.
                ".leading",
                ":leading",
                // Any space at all means somebody already translated it.
                "com.minecolonies.research already translated",
        })
        void notKeys(String input) {
            assertFalse(Text.looksLikeKey(input));
        }
    }

    @Nested
    @DisplayName("stringOrEmpty")
    class StringOrEmpty {

        @Test
        void nullBecomesEmpty() {
            assertEquals("", Text.stringOrEmpty(null));
        }

        @Test
        void valueIsUnchanged() {
            assertEquals("Bob", Text.stringOrEmpty("Bob"));
        }
    }
}

File: common/src/test/kotlin/DahRealPanda/plugins/colonyweb/util/TextTest.kt
package DahRealPanda.plugins.colonyweb.util

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe

/**
 * [Text] turns the identifiers MineColonies hands back through reflection — registry names,
 * blueprint paths, translation keys — into something a dashboard can show a player. It is the
 * one helper whose output is visible on nearly every screen, so its edge cases are worth pinning.
 */
class TextTest : DescribeSpec({

    describe("pathOf") {
        it("drops the namespace of a registry name") {
            Text.pathOf("minecolonies:blockhutbuilder") shouldBe "blockhutbuilder"
        }

        it("returns a name that has no namespace unchanged") {
            Text.pathOf("blockhutbuilder") shouldBe "blockhutbuilder"
        }

        it("keeps everything after the first colon") {
            Text.pathOf("ns:path:with:colons") shouldBe "path:with:colons"
        }

        it("treats a trailing colon as an empty path") {
            Text.pathOf("minecolonies:") shouldBe ""
        }

        it("maps null to the empty string rather than throwing") {
            Text.pathOf(null) shouldBe ""
        }
    }

    describe("humanize") {
        it("title-cases a snake_case identifier") {
            Text.humanize("guard_tower") shouldBe "Guard Tower"
        }

        it("title-cases a kebab-case identifier") {
            Text.humanize("guard-tower") shouldBe "Guard Tower"
        }

        it("splits camelCase on the case change") {
            Text.humanize("guardTower") shouldBe "Guard Tower"
        }

        it("separates letters from digits in both directions") {
            Text.humanize("level2house") shouldBe "Level 2 House"
        }

        it("keeps only the last segment of a path") {
            Text.humanize("blueprints/dev/sawmill1.blueprint") shouldBe "Sawmill 1"
        }

        it("strips a .json extension") {
            Text.humanize("townhall.json") shouldBe "Townhall"
        }

        it("collapses runs of separators instead of emitting empty words") {
            Text.humanize("guard__tower--east") shouldBe "Guard Tower East"
        }

        it("leaves an already-readable name alone") {
            Text.humanize("Guard Tower") shouldBe "Guard Tower"
        }

        it("does not lowercase an acronym it did not split") {
            Text.humanize("NPC_house") shouldBe "NPC House"
        }

        it("returns the empty string for null, empty and blank input") {
            Text.humanize(null) shouldBe ""
            Text.humanize("") shouldBe ""
            Text.humanize("   ") shouldBe ""
        }

        it("returns the empty string when the name is only separators") {
            Text.humanize("___") shouldBe ""
        }
    }

    describe("looksLikeKey") {
        it("accepts a dotted translation key") {
            Text.looksLikeKey("block.minecraft.stone") shouldBe true
        }

        it("accepts a namespaced registry name") {
            Text.looksLikeKey("minecraft:stone") shouldBe true
        }

        it("rejects anything containing a space, because keys never do") {
            Text.looksLikeKey("Guard Tower") shouldBe false
            Text.looksLikeKey("block.minecraft.stone ") shouldBe false
        }

        it("rejects a plain word") {
            Text.looksLikeKey("stone") shouldBe false
        }

        it("rejects a leading separator, which would leave an empty namespace") {
            Text.looksLikeKey(".stone") shouldBe false
            Text.looksLikeKey(":stone") shouldBe false
        }

        it("rejects the empty string") {
            Text.looksLikeKey("") shouldBe false
        }
    }

    describe("stringOrEmpty") {
        it("passes a value through") {
            Text.stringOrEmpty("Guard Tower") shouldBe "Guard Tower"
        }

        it("turns null into the empty string") {
            Text.stringOrEmpty(null) shouldBe ""
        }
    }

    describe("displayName") {
        it("uses a value that is already human-readable") {
            Text.displayName("Guard Tower", "fallback") shouldBe "Guard Tower"
        }

        it("humanizes the last segment of an untranslated key") {
            Text.displayName("block.minecolonies.blockhutbuilder", "fallback") shouldBe "Blockhutbuilder"
        }

        it("drops a .name suffix before humanizing") {
            Text.displayName("item.minecolonies.guard_tower.name", "fallback") shouldBe "Guard Tower"
        }

        it("falls back when the value is null") {
            Text.displayName(null, "fallback") shouldBe "fallback"
        }

        it("falls back when the value is blank") {
            Text.displayName("   ", "fallback") shouldBe "fallback"
        }

        it("falls back when a key humanizes to nothing") {
            Text.displayName("a.__", "fallback") shouldBe "fallback"
        }
    }
})
